joblib/_parallel_backends.py

```diff
@@ -0,0 +1,324 @@
+"""
+Backends for embarrassingly parallel code.
+"""
+
+import gc
+import os
+import sys
+import warnings
+import threading
+import abc
+
+from .format_stack import format_exc
+from .my_exceptions import WorkerInterrupt, TransportableException
+from ._multiprocessing_helpers import mp
+if mp is not None:
+    from .pool import MemmapingPool
+    from multiprocessing.pool import ThreadPool
+
+# Environment variables to protect against bad situations when nesting
+JOBLIB_SPAWNED_PROCESS = "__JOBLIB_SPAWNED_PARALLEL__"
+
+# In seconds, should be big enough to hide multiprocessing dispatching
+# overhead.
+# This settings was found by running benchmarks/bench_auto_batching.py
+# with various parameters on various platforms.
+MIN_IDEAL_BATCH_DURATION = .2
+
+# Should not be too high to avoid stragglers: long jobs running alone
+# on a single worker while other workers have no work to process any more.
+MAX_IDEAL_BATCH_DURATION = 2
+
+
+###############################################################################
+class ParallelBackendBase(object):
+    """Helper abc which defines all methods a ParallelBackend must implement"""
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, parallel):
+        self.parallel = parallel
+
+    @abc.abstractmethod
+    def effective_n_jobs(self, n_jobs):
+        """ Determine the number of jobs which are going to run in parallel """
+        return
+
+    @abc.abstractmethod
+    def initialize(self, n_jobs, poolargs):
+        """ Build a process or thread pool and return the number of workers """
+        return
+
+    @abc.abstractmethod
+    def compute_batch_size(self):
+        """ Determine the optimal batch size """
+        return
+
+    @abc.abstractmethod
+    def batch_completed(self, batch_size, duration):
+        """ Callback indicate how long it took to run a batch """
+        return
+
+    @abc.abstractmethod
+    def terminate(self):
+        """ Shutdown the process or thread pool """
+        return
+
+    @abc.abstractmethod
+    def apply_async(self, func, callback=None):
+        """ Schedule a func to be run """
+        return
+
+    @abc.abstractmethod
+    def get_exceptions(self):
+        return
+
+
+###############################################################################
+class SequentialBackend(ParallelBackendBase):
+    """A ParallelBackend which will execute all batches sequentially.
+    Does not use/create any threading objects, and hence has virtually no
+    overhead. Used when n_jobs == 1."""
+
+    def effective_n_jobs(self, n_jobs):
+        """ Determine the number of jobs which are going to run in parallel """
+        if n_jobs == 0:
+            raise ValueError('n_jobs == 0 in Parallel has no meaning')
+        return 1
+
+    def initialize(self, n_jobs, poolargs):
+        """ Build a process or thread pool and return the number of workers """
+        return 1
```

Comment:
It's bascially telling the Parallel class that we're only able to run 1 task at a time.
However, looking at the actual code in parallel.py, it looks pretty dodgy. 
Returning 1 in any of the backends will cause the Parallel class to switch to the sequential backend to reduce overhead.